src/core/lv_obj_style.c

```diff
@@ -640,6 +642,40 @@ void lv_obj_update_layer_type(lv_obj_t * obj)
     }
 }
 
+
+lv_color32_t lv_obj_get_style_color_filter_resolved(const lv_obj_t * obj, lv_part_t part)
+{
+    const lv_color_filter_dsc_t * f = lv_obj_get_style_color_filter_dsc(obj, part);
+    if(f && f->filter_cb) {
+        lv_opa_t opa = lv_obj_get_style_color_filter_opa(obj, part);
+        lv_color_t color = f->filter_cb(f, lv_color_black(), LV_OPA_COVER);
```

Comment:
It took me 1 hours to understand this line, so I suggest adding a comment here :smile: 
```suggestion
        /*Pass LV_OPA_COVER and any colors. It will return  the pure color that the filter uses to mix the real input color with later*/
        lv_color_t color = f->filter_cb(f, lv_color_black(), LV_OPA_COVER);
```

However, it doesn't work in all cases. The callback can be any complex. E.g. do different things based in the input opacity (< 50% darken, >50% lighten), or based on the color (change the hue).

In light of that I'm not sure if we can cache the color filter's color in the layer.  We can save just a bit indicating if any of the parents had color filter. In most of the cases it will be `0` so we can skip a lot of recursive `style_get` queries.

Reply:
Thanks for your reply, I think I encountered some problems, for an image widget, we have its `recolor`, `recolor_opa`, and `filter_cb`, `filter_opa` styles. we ultimately need to set that to `lv_draw_image_dsc_t`, and I'm not sure how to calculate to get the correct result.

For example, if an image widget’s `recolor` and `recolor_opa` are never set, but it gets `filter_cb` and `filter_opa` from its parent, we cannot get the correct result from `filter_cb` (because it needs a color as a parameter).

Do you have any suggestions for this? Should we consider adding a new feature to implement this function and keep the existing `color_filter`?

Reply:
What if we keep `color_filter` as it is and make `recolor` not only an image but a generic style property, (`style_recolor` and `style_recolor_opa`). It should be not inherited but calculated recursively and cached in the layer (as it's done in the PR).

`color_filter` is still useful as we can do any magic in it but it's more expansive.  We can start using  `recolor` with default theme too and add a config option to enable/disable `color_filter` as it requires a lot of style queries.

In light of that we can keep the `lv_obj_get_style_bg_color_filtered` and similar functions but if `color_filter` is disabled, `lv_obj_style_apply_color_filter` just returns the color as it is. 

We should also add a helper function, like `lv_obj_style_apply_recolor(layer, obj, color)` which is used in `draw_dsc` init functions.

Regarding images, we can just get the global `recolor` and mix it with the `image_recolor`, and start deprecating `image_recolor`.

What do you think?

Reply:
I think I understand what you mean, this is a great idea, I will update this PR later :grinning: